### PR TITLE
feat: whitelist merkle tree

### LINKED ISSUE #12: handle whitelist with a Merkle tree
- [x] add a function in the smart contract to set the merkle root, callable only by owner
- [x] remove mapping storage
- [x] split the buy function in 2 functions, one for the public buy and one for the whitelisted buy. The reason is that they will take different arguments, the whitelisted buy needs the merkle proof
- [x] handle the verification of the merkle proof (see also [cairo-merkle-distributor](https://github.com/ncitron/cairo-merkle-distributor))

Requirements: benchmark data storage costs from other projects to evaluate the opportuniy

Note: If we decide to keep living with the current situation, we may need to add a `remove_from_whitelist` option to keep data as clean as possible and avoid sybil.

Opportunity: By default the storage var of the merkle root will be 0 until an update of the root, we could use this behavior to trigger the fact that the pre-sale is open.

### PR DESCRIPTION
**Pull request by Bal7hazar**
It fixes #12 

- [x] add a function in the smart contract to set the merkle root, callable only by owner
- [x] remove mapping storage
- [x] split the buy function in 2 functions, one for the public buy and one for the whitelisted buy. The reason is that they will take different arguments, the whitelisted buy needs the merkle proof
- [x] handle the verification of the merkle proof

### CODE REVIEW

File: src/interfaces/minter.cairo [hunk 1]
```diff
@@ -33,15 +33,23 @@ namespace ICarbonableMinter:
     func max_supply_for_mint() -> (max_supply_for_mint : Uint256):
     end
 
+    func merkle_root() -> (root : felt):
```
Comment (abdelhamidbakhta) [inline src/interfaces/minter.cairo:36]: I suggest we use a more explicit name like `whitelist_merkle_root`

File: src/mint/library.cairo [hunk 2]
```diff
@@ -296,4 +322,56 @@ namespace CarbonableMinter:
...
     end
+
+    ###
+    # Verify that leaf belongs to merkle tree.
+    # ref: https://github.com/ncitron/cairo-merkle-distributor/blob/master/contracts/merkle.cairo
+    # @param leaf
+    # @param proof_len
+    # @param proof
+    ###
+    func merkle_verify{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
```
Comment (abdelhamidbakhta) [inline src/mint/library.cairo:333]: I suggest we more this function and `calc_merkle_root` in a separate file, can be done in a follow up PR

**Review by abdelhamidbakhta — APPROVED** [review]
left minor comments, otherwise looks good to me

File: src/interfaces/minter.cairo (hunk 1 above)
Comment (Bal7hazar) [reply src/interfaces/minter.cairo]: Done in [98223d1](https://github.com/Carbonable/carbonable-starknet/pull/22/commits/98223d10be7c60a44e759013f91cd29b433bbaeb)

File: src/mint/library.cairo (hunk 2 above)
Comment (Bal7hazar) [reply src/mint/library.cairo]: Done in [98223d1](https://github.com/Carbonable/carbonable-starknet/pull/22/commits/98223d10be7c60a44e759013f91cd29b433bbaeb)

**Pull request merged**